sky/clouds/service_catalog/aws_catalog.py

```diff
@@ -61,25 +70,52 @@ def _apply_az_mapping(df: 'pd.DataFrame') -> 'pd.DataFrame':
     # Use inner join to drop rows with unknown AZ IDs, which are likely
     # because the user does not have access to that Region. Otherwise,
     # there will be rows with NaN in the AvailabilityZone column.
-    df = df.merge(az_mappings, on=['AvailabilityZone'], how='inner')
-    df = df.drop(columns=['AvailabilityZone']).rename(
+    _df = _df.merge(az_mappings, on=['AvailabilityZone'], how='inner')
+    _df = _df.drop(columns=['AvailabilityZone']).rename(
         columns={'AvailabilityZoneName': 'AvailabilityZone'})
-    return df
+    _az_mapping_applied = True
+
+
+def _apply_az_mapping_decorator(func):
+    """Decorator to apply availability zone mapping to the dataframe."""
 
+    def wrapper(*args, **kwargs):
+        with _apply_az_mapping_lock:
+            _apply_az_mapping()
+        return func(*args, **kwargs)
 
-_df = _apply_az_mapping(_df)
+    return wrapper
 
 
+@_apply_az_mapping_decorator
 def instance_type_exists(instance_type: str) -> bool:
     return common.instance_type_exists_impl(_df, instance_type)
 
 
+# Helper functions for `validate_region_zone`.
```

Comment:
Minor preference: I feel like it's clearer to keep these internal to validate_region_zone() (shallow methods). Up to you though.